sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala

```diff
@@ -407,260 +408,284 @@ class GpuSequenceMeta(
 
-        val (sizeWithNegative, diffHasZero) = withResource(stop.sub(start)) { difference =>
-          // sizeWithNegative=floor((stop-start)/step)+1
-          val sizeWithNegative = withResource(difference.floorDiv(step)) { quotient =>
-            quotient.add(one)
-          }
-          val tmpDiffHasZero = closeOnExcept(sizeWithNegative) { _ =>
-            difference.equalTo(zero)
-          }
-          (sizeWithNegative, tmpDiffHasZero)
-        }
+  override def children: Seq[Expression] = Seq(start, stop) ++ stepOpt
 
-        val tmpSize = closeOnExcept(diffHasZero) { _ =>
-          // tmpSize = if(sizeWithNegative < 0) 0 else sizeWithNegative
-          withResource(sizeWithNegative) { _ =>
-            withResource(sizeWithNegative.greaterOrEqualTo(zero)) { pred =>
-              pred.ifElse(sizeWithNegative, zero)
-            }
-          }
-        }
+  override def nullable: Boolean = children.exists(_.nullable)
 
-        // when start==stop && step==0, size will be 0.
-        // but we should change size to 1
-        withResource(tmpSize) { tmpSize =>
-          withResource(diffHasZero) { diffHasZero =>
-            step match {
-              case stepScalar: Scalar =>
-                withResource(ColumnVector.fromScalar(stepScalar, rows)) { stepV =>
-                  withResource(stepV.equalTo(zero)) { stepHasZero =>
-                    withResource(diffHasZero.and(stepHasZero)) { predWithZero =>
-                      predWithZero.ifElse(one, tmpSize)
-                    }
-                  }
-                }
-              case _ =>
-                withResource(step.equalTo(zero)) { stepHasZero =>
-                  withResource(diffHasZero.and(stepHasZero)) { predWithZero =>
-                    predWithZero.ifElse(one, tmpSize)
+  override def foldable: Boolean = children.forall(_.foldable)
+
+  override def columnarEval(batch: ColumnarBatch): Any = {
+    // 1 Compute the array size for all the rows
+    val (startCol, sizeCol, stepCol, nonNullPredOpt) =
+      withResource(columnarEvalToColumn(start, batch)) { startGpuCol =>
+        withResource(columnarEvalToColumn(stop, batch)) { stopGpuCol =>
+          withResource(stepOpt.map(columnarEvalToColumn(_, batch))) { stepColOpt =>
+            // Filter out the nulls before any computation, since cudf 'sequences' does
+            // not allow nulls.(Pls refer to https://github.com/rapidsai/cudf/issues/10012)
+            // Then no need to do computation for the null rows.
+            val (starts, stops, stepsOption, nonNullOption) = filterOutNullsIfNeeded(
+                startGpuCol.getBase, stopGpuCol.getBase, stepColOpt.map(_.getBase))
+
+            // Get the array size of each valid row, along with the steps
+            withResource(stops) { _ =>
+              closeOnExcept(starts) { _ =>
+                closeOnExcept(nonNullOption) { _ =>
+                  closeOnExcept(stepsOption.getOrElse(defaultSteps(starts, stops))) { steps =>
```

Comment:
Do we really want to keep the defaultSteps in the memory?

Reply:
It is not a must, I think.
If there is an improvement, I want to handle it in the following PR.